File: c/timsort.c
#include <stdio.h>
#include <stdlib.h>

#define MIN_MERGE 32
#define MIN_GALLOP 7

void insertionSort(int arr[], int left, int right) {
    for (int i = left + 1; i <= right; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= left && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

void merge(int arr[], int left, int mid, int right) {
    int len1 = mid - left + 1;
    int len2 = right - mid;

    int* leftArr = (int*)malloc(len1 * sizeof(int));
    int* rightArr = (int*)malloc(len2 * sizeof(int));

    for (int i = 0; i < len1; i++)
        leftArr[i] = arr[left + i];
    
    for (int i = 0; i < len2; i++)
        rightArr[i] = arr[mid + i + 1];

    int i = 0, j = 0, k = left;

    while (i < len1 && j < len2) {
        if (leftArr[i] <= rightArr[j])
            arr[k++] = leftArr[i++];
        else
            arr[k++] = rightArr[j++];
    }

    while (i < len1)
        arr[k++] = leftArr[i++];
    
    while (j < len2)
        arr[k++] = rightArr[j++];

    free(leftArr);
    free(rightArr);
}

int minRunLength(int n) {
    int r = 0;
    while (n >= MIN_MERGE) {
        r |= n & 1;
        n >>= 1;
    }
    return n + r;
}

int binarySearch(int arr[], int key, int low, int high) {
    int mid;
    while (low <= high) {
        mid = (low + high) >> 1;
        
        if (arr[mid] < key)
            low = mid + 1;
        else if (arr[mid] > key)
            high = mid - 1;
        else
            return mid;
    }
    return low;
}

void timsort(int arr[], int n) {
    int minRun = minRunLength(MIN_MERGE);
    for (int i = 0; i < n; i += minRun) {
        int left = i;
        int right = (i + minRun - 1 < n) ? (i + minRun - 1) : (n - 1);
        insertionSort(arr, left, right);
    }
    for (int size = minRun; size < n; size <<= 1) {
        for (int left = 0; left < n; left += (size << 1)) {
            int mid = left + size - 1;
            int right = (left + (size << 1) - 1 < n) ? (left + (size << 1) - 1) : (n - 1);
            merge(arr, left, mid, right);
        }
    }
}

int main() {
    int arr[] = {5, 2, 1, 4, 3};
    int n = sizeof(arr) / sizeof(arr[0]);
    
    printf("Before sorting:\n");
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
    
    timsort(arr, n);
    
    printf("After sorting:\n");
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
    
    return 0;
}

File: c/timsort.py
# Определение минимального размера подмассива для сортировки слиянием
MIN_MERGE = 32

def insertion_sort(arr, left, right):
    for i in range(left + 1, right + 1):
        key_item = arr[i]
        j = i - 1
        while j >= left and arr[j] > key_item:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key_item

def merge(arr, left, mid, right):
    len1 = mid - left + 1
    len2 = right - mid
    left_arr = []
    right_arr = []
    for x in range(len1):
        left_arr.append(arr[left + x])
    for x in range(len2):
        right_arr.append(arr[mid + 1 + x])

    i = 0
    j = 0
    k = left
    while i < len1 and j < len2:
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1

    while i < len1:
        arr[k] = left_arr[i]
        k += 1
        i += 1

    while j < len2:
        arr[k] = right_arr[j]
        k += 1
        j += 1

def tim_sort(arr):
    n = len(arr)
    for i in range(0, n, MIN_MERGE):
        insertion_sort(arr, i, min((i + MIN_MERGE - 1), n - 1))

    size = MIN_MERGE
    while size < n:
        for left in range(0, n, 2 * size):
            mid = left + size - 1
            right = min((left + 2 * size - 1), (n - 1))

            merge(arr, left, mid, right)
        size *= 2

# Пример использования:
array = [5, 3, 8, 4, 2, 1, 6, 7]
tim_sort(array)
print(array)
